Build a regression model.

I am creating a simple linear regression model.
#Regression analysis is a series of statistical processes used to estimate the relationships between a dependent variable and various independent variables in statistical modeling. 
#Linear regression seeks the line that best matches the data according to a set of mathematical criteria.
#In simple terms, it uses a straight line to define the relationship between two variables (one independent variable and dependent variable). 
#This is estimating the value of extra.slot variable based on the value of free_slot  variable.



In [1]:
!pip install statsmodels


In [2]:
import statsmodels.api as sm

In [3]:
import numpy as np
import pandas as pd

In [59]:
totalcombined_df= pd.read_csv(r'C:\Users\layin\OneDrive\Desktop\all_python\citybike_folder\file4.csv')



In [60]:
totalcombined_df.columns

Index(['Unnamed: 0', 'empty_slots', 'free_bikes', 'id', 'latitude',
       'longitude', 'name', 'timestamp', 'extra.ebikes', 'extra.has_ebikes',
       'extra.last_updated', 'extra.payment', 'extra.payment-terminal',
       'extra.renting', 'extra.returning', 'extra.slots', 'extra.uid',
       'address', 'category_id', 'category_name', 'postcode', 'distance',
       'review_count', 'rating'],
      dtype='object')

In [61]:
totalcombined_df.shape

(3455, 24)

In [63]:
totalcombined_df.info

<bound method DataFrame.info of       Unnamed: 0  empty_slots  free_bikes                                id  \
0              0          9.0        10.0  553ed0300d38108b4f21a6bafa3db70c   
1              1         31.0         1.0  93542dcbf21f5411569adb92cd7cc199   
2              2         12.0         3.0  d9c7ef5dbda4ed944d1bf51fe540acb6   
3              3         12.0         2.0  0568389e659e679fbe29a5ac12cd49c0   
4              4          4.0        11.0  47b79abc28a54d0e4689b1096ceb8466   
...          ...          ...         ...                               ...   
3450        8723          NaN         NaN                               NaN   
3451        8724          NaN         NaN                               NaN   
3452        8726          NaN         NaN                               NaN   
3453        8728          NaN         NaN                               NaN   
3454        8769          NaN         NaN                               NaN   

       latitude  lo

In [64]:

import statsmodels.api as sm

In [76]:
#A linear regression model using rating and review count as dependent and independent variable.
yelp_df = pd.read_csv(r'C:\Users\layin\OneDrive\Desktop\all_python\3dataframenew\file3.csv')
import pandas as pd
import statsmodels.api as sm

# Load the CSV file into a DataFrame
data = pd.read_csv(r'C:\Users\layin\OneDrive\Desktop\all_python\3dataframenew\file3.csv')
# Select the columns for the analysis
selected_columns =['review_count','distance']

# Convert selected columns to numeric type
data[selected_columns] = data[selected_columns].apply(pd.to_numeric, errors='coerce')
# Drop rows with missing values
data = data.dropna(subset=selected_columns)

# Check if there are enough remaining rows for analysis
if len(data) > 0:
    # Create the feature matrix X and the target variable y
    X = data[selected_columns]
    y = data['rating']

    # Add a constant to the feature matrix
    X = sm.add_constant(X)

    # Perform the linear regression analysis using OLS
    model = sm.OLS(y, X)
    results = model.fit()
    print(results.summary())
else:
    print("Insufficient data for analysis.")

                            OLS Regression Results                            
Dep. Variable:                 rating   R-squared:                       0.013
Model:                            OLS   Adj. R-squared:                  0.013
Method:                 Least Squares   F-statistic:                     57.44
Date:                Mon, 03 Jul 2023   Prob (F-statistic):           1.65e-25
Time:                        14:20:06   Log-Likelihood:                -9967.0
No. Observations:                8793   AIC:                         1.994e+04
Df Residuals:                    8790   BIC:                         1.996e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            3.8460      0.017    221.609   

#the coefficient of distance, has its p-value less than 0.05 so we will remove it. 
The R squared value shows that there is only 13% variation of the rating that can be explained by the linear regression model using the selected features. this implies that this might not be the best model fit to capture a significant portion of the variance on the tergeted variable(rating)..

In [52]:
# A Linear regresion model was built to determine the correllation between extra.slots  of bike and  and free_bikes stations in Boston.
import pandas as pd
import statsmodels.api as sm
import numpy as np

# Load the CSV file into a DataFrame
data2 = pd.read_csv(r'C:\Users\layin\OneDrive\Desktop\all_python\3dataframe folder\file.csv')

# Select the columns for the analysis
selected_columns = ['free_bikes', 'latitude','empty_slots','extra.uid','extra.returning','extra.renting'] # Replace with the actual column names

# Convert selected columns to numeric type
data2[selected_columns] = data2[selected_columns].apply(pd.to_numeric, errors='coerce')

# Drop rows with missing values
data2 = data2.dropna(subset=selected_columns)

# Create the feature matrix X and the target variable y
X = data2[selected_columns]
y = data2['extra.slots']  # Replace 'target_column' with the actual target column name

# Add a constant to the feature matrix
X = sm.add_constant(X)

# Perform the linear regression analysis using OLS
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())


                            OLS Regression Results                            
Dep. Variable:            extra.slots   R-squared:                       0.969
Model:                            OLS   Adj. R-squared:                  0.968
Method:                 Least Squares   F-statistic:                     3483.
Date:                Mon, 03 Jul 2023   Prob (F-statistic):               0.00
Time:                        00:40:26   Log-Likelihood:                -578.32
No. Observations:                 455   AIC:                             1167.
Df Residuals:                     450   BIC:                             1187.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
free_bikes          1.0018      0.011     

Provide model output and an interpretation of the results. 

The R squared indicate that approximately 96.9% of the variation  of the targeted variable( extra.slots of bike stations), can be  explained  by this linear regression model using all this selected features
(free_bikes', 'latitude','empty_slots','extra.uid','extra.returning','extra.renting). This implies that the model fits the data  well and is able to capture a significant portion of variance on the target variable.
Adjusted R  with 96.8% value futher takes into account the no of predictors (features)in the model and adjust the R Squaed value accordingly.
The p-value of latitude  is 0.441,and this p-value  > 0.05 will still be removed from the model, the same goes with  extra.returning and extra.renting with values 0.428.

Overall , all these suggested that there is a strong  significant  relationship between  thefeatures selected  and the targeted variable in this Linear Regression Model.

# Stretch

How can you turn the regression model into a classification model?

To convert a linear regression model into a classification model, I can do the following:

Train your linear regression model using the desired features and target variable.
Obtain the predicted values from the regression model.
Choose a threshold value that will separate the predicted values into classes. This threshold is determined based on analyzing the distribution of predicted values.
Create a new target variable by applying a binary classification based on the threshold. Assign a value of 1 to instances where the predicted value is above the threshold, and a value of 0 to instances below or equal to the threshold.
Train the logistic classification model, using the same features and the new binary target variable.

In [ ]:
#layout of my approach in converting a linear regression into classification model.
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression

# Load the dataset into a DataFrame
data = pd.read_csv(r'C:\Users\layin\OneDrive\Desktop\all_python\3dataframe folder\file.csv')

# Define the features and target variable
features = ['free_bikes', 'latitude','empty_slots','extra.uid','extra.returning','extra.renting']
target = 'extra.slots'

# Train the linear regression model
regression_model = LinearRegression()
regression_model.fit(data[features], data[target])

# Get the predicted values from the regression model
predicted_values = regression_model.predict(data[features])

# Choose a threshold value
threshold = 0.5

# Create a binary target variable based on the threshold
data['binary_target'] = (predicted_values > threshold).astype(int)
#binary_target
# Train a classification model using logistic regression
classification_model = LogisticRegression()
classification_model.fit(data[features], data['binary_target'])
